Build webassembly and windows ports

Josverl opened the pull request:
This pull request includes several changes to add support for building MicroPython for webassembly and windows 

### WebAssembly is supported using the `ARM_BUILD_CONTAINER` :
- The `emsdk` toolchain is installed during the first build using the `tools/ci.sh` script that is used to run the integration tests for MicroPython. 
- after the first build the `emsdk` toolchain is cached and used for subsequent builds.

### Windows is supported using the MinGW toolchain:
- The windows build uses the container definition `micropython/build-micropython-win-mingw`. It uses the same build process as the official MicroPython Docker container, but with the MinGW toolchain instead of the ARM toolchain.
- This container is not yet published, but can be built locally from the `containers/build-micropython-win-mingw` directory

### Build process improvements:

* [`src/mpbuild/build.py`](diffhunk://#diff-6eaf589a9a48d7ebbcba50f9ed067cff680809878db488b64551535f2630b299R25-R26): Added support for the new MinGW build container and updated the build command to handle special cases for `webassembly` and `windows` ports. This includes setting up the CI environment and adding cross-compilation arguments for Windows. [[1]](diffhunk://#diff-6eaf589a9a48d7ebbcba50f9ed067cff680809878db488b64551535f2630b299R25-R26) [[2]](diffhunk://#diff-6eaf589a9a48d7ebbcba50f9ed067cff680809878db488b64551535f2630b299L82-R96) [[3]](diffhunk://#diff-6eaf589a9a48d7ebbcba50f9ed067cff680809878db488b64551535f2630b299R110-R111) [[4]](diffhunk://#diff-6eaf589a9a48d7ebbcba50f9ed067cff680809878db488b64551535f2630b299R129-R130)

### Code cleanup and refactoring:
* [`src/mpbuild/board_database.py`](diffhunk://#diff-c97385d3f93b0bc006eb66b01efd99e2a869b7b512795c2fff8149ceda5ccedeL37-R40): Moved the import of `Path` to the top of the file for consistency and added a print statement to provide more information when a variant is not found. [[1]](diffhunk://#diff-c97385d3f93b0bc006eb66b01efd99e2a869b7b512795c2fff8149ceda5ccedeL37-R40) [[2]](diffhunk://#diff-c97385d3f93b0bc006eb66b01efd99e2a869b7b512795c2fff8149ceda5ccedeR162-R164)

### Depends on 

Never mind - merged already

Josverl committed:
docker_build_cmd: Add support for WebAssembly and Windows builds.

 https://github.com/mattytrentini/mpbuild/issues/13

Josverl committed:
docker_build_cmd: Add support for WebAssembly and Windows builds.

 https://github.com/mattytrentini/mpbuild/issues/13

Josverl commented:
@mattytrentini , in order to build windows a new container needs to be published.
is that something that you can do ? 
I'd rather not scatter the build images across different owners, but not sue if to host them in seperate repos ?

mattytrentini commented:
> is that something that you can do ?
> I'd rather not scatter the build images across different owners, but not sue if to host them in seperate repos ?

Yes, I can do that (I can push to the MicroPython docker hub) but I would prefer to have the Dockerfile's in their own separate repositories. Currently I'm the owner for the repositories of containers to create builds (that aren't published elsewhere, like the IDF) so I'm happy to create a new repo to house the mingw build container you've listed, if that's ok with you?

Long-term, I'd like to transfer those repos to the micropython Github org.

Josverl commented:
> I'm happy to create a new repo to house the mingw build container you've listed, if that's ok with you?

Perfect, if its in place i'll splice the PR

andrewleech commented:
I'm not sure if it's faster or not but there's an official emsdk image that can be used: 
```
docker run -it --rm -v $(pwd):$(pwd) -w $(pwd) emscripten/emsdk bash -c 'source tools/ci.sh && ci_webassembly_setup && ci_webassembly_build && ci_webassembly_run_tests'
```

Josverl commented:
From the first look That are essentially the same commands, running the CI setup scripts.

The value of Embedding them in Mpflash is that it makes building more accessible.

But I agree  that the logic structure for stringing together the docker commands could use improvement

Josverl commented:
It's useful to me for sure 😊

Josverl committed:
Merge branch 'main' into build_sa_ports

Josverl committed:
Merge branch 'main' into build_sa_ports

mattytrentini committed:
Fix CI: lint, format, and webassembly-now-supported test

After PR #70 merged main back into the branch, the modernised
codebase (PR #102's ruff config — E, F, I, UP, B; line-length=100)
flagged six pre-existing issues in this PR's diff:

- I001  unsorted imports across the merge boundary
- F811  `from pathlib import Path` duplicated (added on a new line
        and the original kept too)
- UP035 `from typing import List, Optional` — replaced by stdlib
        generics in PR #102
- F401  List/Optional imported but unused
- E501  the `build_container = (...)` ternary collapsed to a single
        line by the merge, exceeding 100 chars
- A formatter difference: this PR used single-quoted f-strings for
  the ci_*_cmd interpolations; the rest of the file uses double.

`uv run ruff check --fix` + `uv run ruff format` resolves all six.

The single test failure was real: `test_unsupported_port_raises` in
test_build_container.py used `webassembly` as the canonical
unsupported port, but this PR makes webassembly supported (with
ARM_BUILD_CONTAINER + a tools/ci.sh emsdk install). Switched the
test to a synthesised `qemu` port that genuinely isn't in
BUILD_CONTAINERS, and added two new tests covering the now-supported
`webassembly` and `windows` special-port resolutions.

All checks now green locally:
- uv run ruff check / format --check
- uv run ty check
- uv run pytest      (101 passed)
- uv run pre-commit run --all-files

mattytrentini committed:
Fix CI: lint, format, and webassembly-now-supported test

After PR #70 merged main back into the branch, the modernised
codebase (PR #102's ruff config — E, F, I, UP, B; line-length=100)
flagged six pre-existing issues in this PR's diff:

- I001  unsorted imports across the merge boundary
- F811  `from pathlib import Path` duplicated (added on a new line
        and the original kept too)
- UP035 `from typing import List, Optional` — replaced by stdlib
        generics in PR #102
- F401  List/Optional imported but unused
- E501  the `build_container = (...)` ternary collapsed to a single
        line by the merge, exceeding 100 chars
- A formatter difference: this PR used single-quoted f-strings for
  the ci_*_cmd interpolations; the rest of the file uses double.

`uv run ruff check --fix` + `uv run ruff format` resolves all six.

The single test failure was real: `test_unsupported_port_raises` in
test_build_container.py used `webassembly` as the canonical
unsupported port, but this PR makes webassembly supported (with
ARM_BUILD_CONTAINER + a tools/ci.sh emsdk install). Switched the
test to a synthesised `qemu` port that genuinely isn't in
BUILD_CONTAINERS, and added two new tests covering the now-supported
`webassembly` and `windows` special-port resolutions.

All checks now green locally:
- uv run ruff check / format --check
- uv run ty check
- uv run pytest      (101 passed)
- uv run pre-commit run --all-files

mattytrentini commented:
OK @Josverl, [micropython/build-micropython-win-mingw:2.0](https://hub.docker.com/r/micropython/build-micropython-win-mingw) has been published. Note that it was `just build push` (not _publish_) that was required. 😛 Thanks for those scripts though, they were handy!

Josverl commented:
Thanks Matt,

Yes the just files are quite useful for simple things, and avoid a clutter of scripts.
